### PR TITLE
Add basic HMR support for Svelte 3 (and svelte-native)

### PR DESCRIPTION
**Pull request by rixo**
This is a rework of the experimental code for Svelte 3 HMR I've presented in https://github.com/sveltejs/svelte-loader/issues/92.

The loader can be tried with [this demo app](https://github.com/rixo/demo-svelte3-hmr).

The rework builds around existing code instead of replacing it, in order to preserve existing support for svelte 2 + HMR. Also, this offers a more readable diff, as suggested by @ekhaled in https://github.com/sveltejs/svelte/pull/3126#issuecomment-506800833.

Actual HMR support works for basic cases. Not all cases unfortunately... Outros, for example, are known to break HMR currently. I am working in parallel on a test suite for HMR, to identify, fix & consolidate such cases.

But my goal here is more to get the ball rolling, foster some discussion and, hopefully, some decisions...

What would be great would be to be able to put version numbers on this ongoing work soon, so that people can pin versions of their stack when it works. This is not possible when installing the package from a git branch, and it is a pity because Svelte is evolving fast, under-development HMR will break often... And sometimes people just want to work on their actual project, not on fixing their HMR (or worse: loader) that was working before.

This implementation emulates svelte's `capture_state` and `inject_state` hooks as proposed in https://github.com/sveltejs/svelte/issues/2377. My hope is that this work will usefully inform the actual implementation of HMR hooks in Svelte, when time comes. So I've moved all the emulated code / code that touches svelte private API into their [own file](https://github.com/rixo/svelte-loader/blob/hmr/lib/svelte3/hot/svelte-hooks.js) to help with that.

### CODE REVIEW

File: lib/resolve-svelte.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+			base: 'svelte',
+		};
+	}
+};
+
+const { req, base } = resolveSvelte();
+
+const { version } = req(`${base}/package.json`);
+
+const major_version = +version[0];
```
Comment (DominusVilicus) [inline lib/resolve-svelte.js:26]:
```suggestion
const [major_version, minor_version ]= version;
```

File: lib/resolve-svelte.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+const { compile, preprocess } =
+	major_version >= 3 ? req(`${base}/compiler`) : req(`${base}`);
+
+const makeHot =
+	major_version >= 3 ? require('./svelte3/make-hot') : require('./make-hot');
+
+module.exports = {
+	major_version,
```
Comment (DominusVilicus) [inline lib/resolve-svelte.js:35]:
```suggestion
	major_version,
	minor_version,
```

File: index.js [hunk 3]
```diff
@@ -1,14 +1,14 @@
...
 
-const hotApi = require.resolve('./lib/hot-api.js');
-
-const { version } = require('svelte/package.json');
-const major_version = +version[0];
-const { compile, preprocess } = major_version >= 3
-	? require('svelte/compiler')
-	: require('svelte');
+const {
+	major_version,
```
Comment (DominusVilicus) [inline index.js:7]:
```suggestion
	major_version,
	minor_version,
```

**Comment by DominusVilicus** [comment]
What's causing the tests to fail?

**Comment by rixo** [comment]
@DominusVilicus I think I incorrectly removed svelte from package.json...

Anyway, I don't think this PR is current anymore. HMR implementation has changed a lot and integrating will need a fresh look. Closing this one to reduce confusion.

My goal is to keep `svelte-loader-hot` in sync with official `svelte-loader`, so if your change gets pulled here, I'll merge it in `svelte-loader-hot` as well (feel free to PR when this happens!).

**Comment by DominusVilicus** [comment]
No probs. Mention me when you add the new HMR implementation in a PR, would be happy to review.

**Comment by rixo** [comment]
It is not really planned to integrate HMR into the official plugins short term. For now the plan with HMR is increase adoption, get confidence, stabilize and then make it official.

That being said, the plan is also to keep the hot plugins in sync with official ones. So like I said, anything that makes it here will be merged in svelte-loader-hot too (feel free to PR when you feel this process is lagging, I'm a bit behind tracking all this stuff ^^).

**Pull request closed without merging**